### PR TITLE
Fix PD decode hang with DP attention and GLM-5.2 MTP

### PR DESCRIPTION
**Pull request by HZY-Wade**
## Motivation

A PD-disaggregated nvidia/GLM-5.2-NVFP4 deployment can hang or crash after decode requests when DP attention and EAGLE MTP are enabled together. The equivalent non-disaggregated deployment completes the same requests.

This is a follow-up to #30839. A minimal reproduction uses the following relevant arguments. Mooncake and RDMA arguments may need adjustment for the local network.

Prefill:

    python -m sglang.launch_server \
      --model-path nvidia/GLM-5.2-NVFP4 \
      --trust-remote-code \
      --quantization modelopt_fp4 \
      --moe-runner-backend flashinfer_trtllm \
      --disaggregation-mode prefill \
      --disaggregation-bootstrap-port 8998 \
      --disaggregation-transfer-backend mooncake \
      --tp 8 \
      --dsa-prefill-backend trtllm \
      --host 0.0.0.0 \
      --port 30000

Decode:

    python -m sglang.launch_server \
      --model-path nvidia/GLM-5.2-NVFP4 \
      --trust-remote-code \
      --quantization modelopt_fp4 \
      --moe-runner-backend flashinfer_trtllm \
      --disaggregation-mode decode \
      --disaggregation-bootstrap-port 8998 \
      --disaggregation-transfer-backend mooncake \
      --tp 8 \
      --dp-size 8 \
      --enable-dp-attention \
      --dsa-decode-backend trtllm \
      --speculative-algorithm EAGLE \
      --speculative-num-steps 3 \
      --speculative-eagle-topk 1 \
      --speculative-num-draft-tokens 4 \
      --host 0.0.0.0 \
      --port 30001

Five related failure modes or follow-up issues were observed.

First, the rank receiving the real request can enter the eager MTP draft path while idle DP ranks replay the target verification CUDA graph. The ranks then reach different collective phases and the scheduler eventually triggers the watchdog.

For GLM-5.2 IndexShare, a PD decode round can arrive without dsa_topk_indices. The old fallback disabled graph replay only on the active rank. Idle ranks remained graph-eligible even though all ranks participate in the same DP and EP forward.

Second, overlap scheduling can make the graph vote observe stale DSA seed state. When a seeded running batch merges a seedless prebuilt batch, future_dsa_topk_indices_available becomes false while the materialized dsa_topk_indices tensor can remain non-empty. The future flag must be authoritative whenever future inputs exist, otherwise the active rank falls back only after the DP-wide vote.

Third, after synchronizing eager fallback, DP padding exposed an eager draft postprocessing mismatch:

    draft_topk1_postprocess
      assert positions.shape[0] == next_token_logits.shape[0]

EAGLE postprocessing must consume only rank-local token rows rather than padded logits.

Fourth, idle eager ranks and active padded ranks exposed two DSA shape mismatches. Idle ranks do not have real page-table rows and must skip sparse index selection. Active ranks must trim padded query and top-k rows to the real batch described by pre-planned metadata, run attention on that prefix, then restore the physical DP shape before later MLP and EP collectives.

Fifth, the correctness-first fix originally folded seedless fallback into the generic DP graph gate. This kept ranks consistent but also forced target verification and draft-extend eager whenever only the EAGLE draft phase lacked a DSA seed. Under the reproduced workload, about 9.7% of reported decode batches lost generic CUDA graph replay, creating an avoidable performance regression.

## Modifications

- Add a speculative-worker hook that reports when a batch requires rank-consistent eager draft execution.
- Detect the GLM-5.2 seedless IndexShare condition before DP MLP synchronization in disaggregated decode.
- Treat future_dsa_topk_indices_available as authoritative under overlap scheduling.
- Append a dedicated draft-graph eligibility vote to the existing DP MLP-sync all-gather. No additional collective is introduced.
- Make EAGLEDraftCudaGraphRunner consume both the generic graph vote and the draft-specific vote.
- Keep target verification and draft-extend on their normal CUDA graph path when only draft seed state is missing.
- Slice eager draft logits, hidden states, and positions to the local draft-token count.
- Short-circuit eager DSA index selection on idle DP ranks with all-invalid top-k rows.
- Trim TRT-LLM DSA attention inputs to the real metadata batch and zero-pad the output back to the physical DP shape.
- Add regression coverage for current and future seed states, overlap batch merging, independent generic and draft graph votes, active and idle DP padding, DSA idle indexing, and TRT-LLM page-table alignment.

## Accuracy and Stability Tests

The changes synchronize control flow and isolate DP padding from metadata that describes only real request rows. Model math for real tokens is unchanged.

- Repository pre-commit hooks pass for all modified files.
- Focused local suite: 14 tests passed, plus 4 subtests.
- The patch was hot-applied to the reproduced PD decode deployment with TP8, DP8, DP attention, TRT-LLM DSA, NVFP4, and EAGLE MTP.
- Under mirrored and validation traffic, all 8 scheduler ranks remained alive.
- 2,019 reported decode batches completed with generic CUDA graph enabled: 2,019 true and 0 false.
- Per-rank running batch size reached 7 during observation.
- No scheduler exception, assertion, watchdog, or CUDA graph collective hang occurred.

## Speed Tests and Profiling

The draft-specific gate preserves the required rank-consistent eager fallback but limits it to the EAGLE draft phase.

Before this follow-up, approximately 9.7% of reported decode batches disabled the generic graph path because a seedless draft also disabled target verification and draft-extend graphs. After this follow-up, the observed deployment reported 0 generic graph fallbacks across 2,019 decode batch samples.

This is an operational validation rather than a controlled end-to-end throughput benchmark. The expected gain is workload-dependent and proportional to the frequency of seedless draft rounds; no extra DP collective is added.

## Checklist

- [x] Format the code with repository pre-commit hooks.
- [x] Add unit regression tests.
- [x] Follow the SGLang code style guidance.
- [x] Validate on the reproduced PD + DP attention + EAGLE deployment.
- [ ] Documentation update (not applicable; no user-facing API change).
- [ ] Controlled accuracy and speed benchmark tables.

---
### CI States

Latest PR Test (Base): :x: [Run #30020688468](https://github.com/sgl-project/sglang/actions/runs/30020688468)
Latest PR Test (Extra): :x: [Run #30020688547](https://github.com/sgl-project/sglang/actions/runs/30020688547)

### CODE REVIEW

**Comment by HZY-Wade** [comment]
@ShangmingCai cc

**Comment by IvanShan177** [comment]
Independent reproduction and corroboration of this failure mode, from a deployment that differs from yours in two ways that may be useful to you.

**The transfer backend is different, and it does not matter.** We run `--disaggregation-transfer-backend nixl`, not mooncake, and see the same hang. So the bug is not in the KV transport.

**Bisected to #30839 by image lineage.** Four otherwise-identical runs of the same GLM-5.2-NVFP4 PD recipe (2P2D, DP attention on, EAGLE MTP on), varying only the container image:

| nightly image (source date) | contains #30839 | first routed request | result |
|---|---|---|---|
| `20260711-7de33ce8` (07-11) | no | 9.609 s, correct answer, `spec_accept_length` 3.24 | pass |
| `20260713-b94ac87e` (07-12) | no | 7.822 s, correct answer, `spec_accept_length` 3.24 | pass |
| `20260714-2f79d334` (07-13) | no | 6.317 s, correct answer, `spec_accept_length` 3.24 | pass |
| `20260715-50d1edaa` (07-14) | **yes** | 240 s timeout, no response | **hang** |

A targeted revert of #30839 inside the newer image also restores serving, and a same-box A/B against `v0.5.15.post1` (which carries the release cherry-pick #31083) reproduces the hang while the pre-#30839 nightly serves the identical request in 3.4 s.

**Symptom detail, in case it helps narrow the collective divergence.** Readiness and health endpoints stay green and both P/D warmups return 200. The first routed generation then never returns. Meanwhile:

- decode GPUs sit at 100% utilization,
- every decode rank reports `num_running_reqs = 0`, `token_usage = 0`, `spec_accept_length = 0.0` (252/252 samples),
- `generation_tokens_total` stays frozen at its warmup value,
- and **no log signature appears at all** — no `KVPoll.WaitingForInput`, no `Prefill bootstrap failed`, no traceback, no OOM.

The complete absence of log evidence while the GPUs spin is consistent with your first failure mode (ranks stuck in different collective phases).

**Not hardware- or shape-specific.** Reproduced on both GB300 and B300 hosts, and at both 4P4D and 2P2D shapes.

**Aggregated deployments are unaffected**, matching your observation — the same image serves correctly in a non-disaggregated TP8 configuration with EAGLE MTP on (`spec_accept_length` 4.7).

**A zero-code mitigation that also corroborates your root cause.** Setting `index_share_for_mtp_iteration = false` (we pass it via `--json-model-override-args`) fully restores serving: 3.446 s versus 3.439 s on the pre-#30839 image, with `spec_accept_length` 3.239, i.e. speculative decoding still healthy. Notably `SGLANG_DSA_FUSE_TOPK=false` does **not** help, which points at the IndexShare seeding path itself rather than fused top-k.

This may be worth mentioning as an operator workaround for anyone who cannot take this patch immediately.

The PR Test runs on this PR are currently red. If it would help, we have a deterministic reproduction — the first routed request hangs every time — and can hot-apply the patch there and report back.

**Comment by IvanShan177** [comment]
One more observation, offered as a test-coverage gap rather than a criticism — it may explain how this reached a stable release.

`#30839` added 96 lines to `test/srt/disaggregation/test_disaggregation_wire.py`, but none of those lines enable speculative decoding, EAGLE, MTP, or DP attention. More broadly, at `v0.5.15.post1` the entire `test/srt/disaggregation/` directory contains **zero** references to `speculative` or `eagle`.

So the PR whose subject was *"MTP IndexShare across PD"* extended the disaggregation wire test without ever turning MTP on, and the disaggregation suite as a whole never exercises the PD + spec combination. That is the axis this hang lives on.

The encouraging part is how cheap a guard would be. The hang is **deterministic on the very first routed request** — no soak, no load, no concurrency needed. A single PD pair with `--enable-dp-attention` and `--speculative-algorithm EAGLE`, one request, and a timeout is sufficient. We measured 3–10 s for a healthy first response versus a hang that never returns.

The obvious obstacle is that this needs multi-GPU CI with a live prefill/decode pair, which is presumably why it does not exist today. If a pre-merge signal on this PR would help, the offer in our previous comment stands: we have a deterministic reproduction on GB300 and B300 hardware and can run an integration check for you.

We are also wiring the equivalent gate into our own pipeline, because our validation had the same blind spot — every one of our passing PD runs this week had speculative decoding disabled for memory reasons, which is exactly why we did not catch this until we finally ran a spec-enabled PD shape.

**Comment by HZY-Wade** [comment]
Thanks for digging into the test-coverage gap and for offering deterministic multi-GPU validation. I agree that the focused unit tests in this PR cover the control-flow and padding cases, but they cannot replace a live PD + DP attention + EAGLE integration test. A hot-applied check on your GB300/B300 setup would be very helpful; please share the launch arguments and result here.

One additional deployment observation from our reproduction: enabling MTP on the prefill side also avoids the hang. In that configuration, the IndexShare seed / `dsa_topk_indices` is consistently available when decode starts, so we do not enter the seedless, rank-divergent draft fallback that triggers the deadlock. We view this as an operator workaround and corroborating signal rather than the correctness fix: prefill-side MTP should not be required for a decode deployment that enables EAGLE MTP, and seedless arrivals still need rank-consistent handling.

I agree that a dedicated multi-GPU regression covering PD + DP attention + EAGLE, including the prefill-without-MTP case, would be the right long-term guard. Thanks again for also wiring this into your pipeline.

**Comment by IvanShan177** [comment]
**Correction to my previous comment — I got that wrong, and the author deserves the record set straight.**

I cited the path as `test/srt/disaggregation/` and claimed the directory contained zero references to `speculative` or `eagle`. Both parts were wrong. The real path is `test/registered/unit/disaggregation/`, and my "zero references" result came from grepping a directory that does not exist — an empty result I mistook for evidence. Apologies for the noise.

The accurate picture is close to the opposite of what I wrote. `test_disaggregation_wire.py` does reference speculative/EAGLE/MTP, and of the 96 lines `#30839` added, ten touch exactly that area. The tests it added are specifically about the seed path this bug lives on:

- `test_metadata_buffer_copies_seed_and_uses_invalid_sentinel`
- `test_decode_input_requires_valid_seed_for_every_request`
- `test_future_map_initializes_seed_buffer_after_seedless_payload`

So `#30839` did add seed-aware coverage for the seedless IndexShare case. My criticism was unfounded.

**The real gap is narrower and, I think, more interesting.** Those are unit tests over the seed plumbing on a single logical path. The failure in this PR is a *multi-rank runtime disagreement* — the rank holding the real request goes eager while idle DP ranks replay the target-verification CUDA graph, so the ranks land in different collective phases. No unit test over seed buffers can observe that, because the divergence only exists once several ranks are running the same forward with different graph eligibility.

That is an integration-level axis: a live prefill/decode pair, `--enable-dp-attention`, `--speculative-algorithm EAGLE`, and at least two DP ranks so that one can be idle while another is active. The cheap part still holds — the hang is deterministic on the first routed request, so one request and a timeout is enough; healthy first response was 3–10 s in our runs versus a hang that never returns. The expensive part is that it needs multi-GPU CI, which is presumably why this axis is not covered.

Our offer stands and is unchanged: we have a deterministic reproduction on GB300 and B300 and are happy to run an integration check for this PR if a pre-merge signal is useful.

**Comment by IvanShan177** [comment]
I opened draft regression test #32722 for GLM-5.2 PD disaggregation + DP attention + MTP; it is intentionally RED on current `main` pending #32209, with the four-GPU TP2+TP2 memory-fit prerequisite explicitly left unverified in the PR.

**Comment by HZY-Wade** [comment]
Hi @IvanShan177 , just checking whether you are still available to review this PR when you have time. AMD-AGI/Infera#58 has now independently reproduced the same issue on ROCm and adopted the core DP draft-graph vote from this PR. Please let me know if I should ask another reviewer instead. Thanks!
